File: tests/src/test/java/com/github/klee0kai/stone/test/inject/InjectWrappersTests.java
package com.github.klee0kai.stone.test.inject;

import com.github.klee0kai.test.mowgli.Forest;
import com.github.klee0kai.test.mowgli.animal.Mowgli;
import org.junit.jupiter.api.Test;

import java.util.Objects;

import static org.junit.jupiter.api.Assertions.*;

public class InjectWrappersTests {

    @Test
    public void supportWrappersTest() {
        //Given
        Forest forest = new Forest();
        forest.create();

        //When
        Mowgli mowgli = new Mowgli();
        mowgli.born();

        //Then
        assertNotNull(mowgli.knowledge);
        assertNotNull(mowgli.knowledgeWeakRef.get());
        assertNotNull(mowgli.knowledgeSoftRef.get());
        assertNotNull(mowgli.knowledgeLazyProvide.get());
        assertNotNull(mowgli.knowledgePhantomProvide.get());
        assertNotNull(mowgli.knowledgePhantomProvide2.get());
        assertNotNull(mowgli.knowledgePhantomProvide3.get());
    }

    @Test
    public void refWrapperTest() {
        //Given
        Forest forest = new Forest();
        forest.create();

        //When
        Mowgli mowgli = new Mowgli();
        mowgli.born();

        //Then
        assertEquals(
                Objects.requireNonNull(mowgli.knowledgeWeakRef.get()).uuid,
                Objects.requireNonNull(mowgli.knowledgeWeakRef.get()).uuid
        );
        assertEquals(
                Objects.requireNonNull(mowgli.knowledgeSoftRef.get()).uuid,
                Objects.requireNonNull(mowgli.knowledgeSoftRef.get()).uuid
        );
        assertEquals(
                Objects.requireNonNull(mowgli.knowledgeLazyProvide.get()).uuid,
                Objects.requireNonNull(mowgli.knowledgeLazyProvide.get()).uuid
        );
    }

    @Test
    public void genWrapperTest() {
        //Given
        Forest forest = new Forest();
        forest.create();

        //When
        Mowgli mowgli = new Mowgli();
        mowgli.born();

        //Then
        assertNotEquals(
                mowgli.knowledgePhantomProvide.get().uuid,
                mowgli.knowledgePhantomProvide.get().uuid
        );
        assertNotEquals(
                mowgli.knowledgePhantomProvide2.get().uuid,
                mowgli.knowledgePhantomProvide2.get().uuid
        );
        assertNotEquals(
                mowgli.knowledgePhantomProvide3.get().uuid,
                mowgli.knowledgePhantomProvide3.get().uuid
        );
    }
}

File: test_kotlin/src/test/kotlin/com/github/klee0kai/kotlin_test/inject/InjectWrappersTests.kt
package com.github.klee0kai.kotlin_test.inject

import com.github.klee0kai.test_kotlin.mowgli.RainForest
import com.github.klee0kai.test_kotlin.mowgli.animal.Gorilla
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class InjectWrappersTests {

    @Test
    fun supportWrappersTest() {
        //Given
        val forest = RainForest()
        forest.create()

        //When
        val gorilla = Gorilla()
        gorilla.born()

        //Then
        Assertions.assertNotNull(gorilla.knowledge)
        Assertions.assertNotNull(gorilla.knowledgeWeakRef.get())
        Assertions.assertNotNull(gorilla.knowledgeSoftRef.get())
        Assertions.assertNotNull(gorilla.knowledgeLazyProvide!!.get())
        Assertions.assertNotNull(gorilla.knowledgePhantomProvide!!.get())
        Assertions.assertNotNull(gorilla.knowledgePhantomProvide2!!.get())
        Assertions.assertNotNull(gorilla.knowledgePhantomProvide3!!.get())
    }


    @Test
    fun refWrapperTest() {
        //Given
        val forest = RainForest()
        forest.create()

        //When
        val gorilla = Gorilla()
        gorilla.born()


        //Then
        Assertions.assertEquals(
            gorilla.knowledgeWeakRef.get()!!.uuid,
            gorilla.knowledgeWeakRef.get()!!.uuid
        )
        Assertions.assertEquals(
            gorilla.knowledgeSoftRef.get()!!.uuid,
            gorilla.knowledgeSoftRef.get()!!.uuid
        )
        Assertions.assertEquals(
            gorilla.knowledgeLazyProvide!!.get()!!.uuid,
            gorilla.knowledgeLazyProvide!!.get()!!.uuid
        )
    }

    @Test
    fun genWrapperTest() {
        //Given
        val forest = RainForest()
        forest.create()

        //When
        val gorilla = Gorilla()
        gorilla.born()

        //Then
        Assertions.assertNotEquals(
            gorilla.knowledgePhantomProvide!!.get().uuid,
            gorilla.knowledgePhantomProvide!!.get().uuid
        )
        Assertions.assertNotEquals(
            gorilla.knowledgePhantomProvide2!!.get().uuid,
            gorilla.knowledgePhantomProvide2!!.get().uuid
        )
        Assertions.assertNotEquals(
            gorilla.knowledgePhantomProvide3!!.get().uuid,
            gorilla.knowledgePhantomProvide3!!.get().uuid
        )
    }

}